In [ ]:
# ==========================================================
# IMPORTAÇÃO DAS BIBLIOTECAS
# ==========================================================

# pandas -> leitura e manipulação do dataset
import pandas as pd

# math -> utilizar a raiz quadrada
import math

# numpy -> Cálculos Mametáticos
import numpy as np

# Counter -> contar qual classe aparece mais vezes
from collections import Counter

# Métricas para avaliação do modelo
from sklearn.metrics import classification_report
from sklearn.metrics import confusion_matrix

In [ ]:
# ==========================================================
# LEITURA DO DATASET
# ==========================================================

# Carrega o arquivo CSV
df = pd.read_csv("https://drive.google.com/uc?export=download&id=1wF7amDwgNKIxa9Lc5Dm4OXkiDy_N_CYu")

# Embaralha os dados
df = df.sample(frac=1, random_state=42).reset_index(drop=True)

# Mostra as 5 primeiras linhas
print(df.head())

# Mostra quantidade de linhas e colunas
print("\nFormato do dataset:", df.shape)


   battery_power  bluetooth  clock_speed  dual_sim  front_camera  four_g  \
0           1392          0          2.5         0             0       1   
1           1503          1          0.7         0            10       1   
2            765          0          2.9         0             0       1   
3            754          0          0.5         1             7       1   
4           1811          1          2.5         0             4       1   

   storage  mobile_weight  n_cores  primary_camera  px_height  px_width   ram  \
0       44            113        4               2        482      1098  1280   
1       43            166        4              14        841      1304  2122   
2       18            153        7               0          6       793  1066   
3       59            178        7              10       1914      1928  1027   
4        5             98        4              16        447       568  2700   

   sc_height  sc_width  talk_time  three_g  touch_screen

In [ ]:
# ==========================================================
# SEPARAÇÃO DOS ATRIBUTOS E DA VARIÁVEL ALVO
# ==========================================================

# Colunas escolhidas para o KNN
colunas_usadas = [
    "battery_power",
    "clock_speed",
    "n_cores",
    "ram"
]

# X -> atributos utilizados
X = df[colunas_usadas].values

# y -> variável alvo
y = df["price_range"].values

In [ ]:
# ==========================================================
# NORMALIZAÇÃO MANUAL DOS DADOS
# ==========================================================

# Converte para array NumPy
X = np.array(X, dtype=float)

# Calcula mínimos e máximos
minimos = X.min(axis=0)
maximos = X.max(axis=0)

# Aplica normalização Min-Max
X = (X - minimos) / (maximos - minimos)

print("\nDados normalizados com sucesso!")


Dados normalizados com sucesso!


In [ ]:
# ==========================================================
# DIVISÃO ENTRE TREINO E TESTE
# ==========================================================

# Utiliza 80% para treino
tamanho_treino = int(0.8 * len(X))

# Dados de treino
X_treino = X[:tamanho_treino]
y_treino = y[:tamanho_treino]

# Dados de teste
X_teste = X[tamanho_treino:]
y_teste = y[tamanho_treino:]

print("\nQuantidade treino:", len(X_treino))
print("Quantidade teste:", len(X_teste))


Quantidade treino: 1598
Quantidade teste: 400


In [ ]:
# ==========================================================
# FUNÇÃO DA DISTÂNCIA EUCLIDIANA
# ==========================================================

# Calcula a distância entre dois pontos
def distancia_euclidiana(ponto1, ponto2):

    diferenca = ponto1 - ponto2

    distancia = np.sqrt(np.sum(diferenca ** 2))

    return distancia

In [ ]:
# ==========================================================
# FUNÇÃO PRINCIPAL DO KNN
# ==========================================================

def knn_prever(X_treino, y_treino, novo_ponto, k):

    # Lista para armazenar distâncias
    distancias = []

    # ======================================================
    # CALCULAR DISTÂNCIA ENTRE O NOVO PONTO
    # E TODOS OS DADOS DE TREINO
    # ======================================================

    for i in range(len(X_treino)):

        distancia = distancia_euclidiana(
            novo_ponto,
            X_treino[i]
        )

        classe = y_treino[i]

        # Armazena distância e classe
        distancias.append((distancia, classe))


    # ======================================================
    # ORDENAR DA MENOR DISTÂNCIA PARA A MAIOR
    # ======================================================

    distancias.sort(key=lambda item: item[0])


    # ======================================================
    # PEGAR OS K VIZINHOS MAIS PRÓXIMOS
    # ======================================================

    vizinhos = distancias[:k]


    # ======================================================
    # PEGAR AS CLASSES DOS VIZINHOS
    # ======================================================

    classes_vizinhos = []

    for vizinho in vizinhos:
        classes_vizinhos.append(vizinho[1])


    # ======================================================
    # DESCOBRIR QUAL CLASSE APARECE MAIS
    # ======================================================

    classe_prevista = Counter(classes_vizinhos).most_common(1)[0][0]

    return classe_prevista

In [ ]:
# ==========================================================
# ESCOLHA DO VALOR DE K
# ==========================================================

# Quantidade de vizinhos
k = 65

print("\nValor escolhido para K:", k)


Valor escolhido para K: 65


In [ ]:
# ==========================================================
# REALIZAR PREVISÕES
# ==========================================================

previsoes = []

# Percorre todos os dados de teste
for ponto in X_teste:

    # Faz previsão
    classe_prevista = knn_prever(
        X_treino,
        y_treino,
        ponto,
        k
    )

    # Salva previsão
    previsoes.append(classe_prevista)

In [ ]:
# ==========================================================
# CALCULAR ACURÁCIA
# ==========================================================

acertos = 0

# Verifica quantas previsões acertaram
for i in range(len(y_teste)):

    if previsoes[i] == y_teste[i]:
        acertos += 1

# Fórmula da acurácia
acuracia = acertos / len(y_teste)

In [ ]:
# ==========================================================
# RESULTADOS FINAIS
# ==========================================================

print("\n================ RESULTADOS ================")

print("Total de testes:", len(y_teste))

print("Total de acertos:", acertos)

print("Acurácia:", acuracia)

print("Acurácia em porcentagem:",
      round(acuracia * 100, 2), "%")


================ RESULTADOS ================
Total de testes: 400
Total de acertos: 326
Acurácia: 0.815
Acurácia em porcentagem: 81.5 %


In [ ]:
# ==========================================================
# CLASSIFICATION REPORT
# ==========================================================

print("\n================ CLASSIFICATION REPORT ================")

print(classification_report(y_teste, previsoes))


# ==========================================================
# MATRIZ DE CONFUSÃO
# ==========================================================

print("\n================ MATRIZ DE CONFUSÃO ================")

print(confusion_matrix(y_teste, previsoes))


================ CLASSIFICATION REPORT ================
              precision    recall  f1-score   support

           0       0.83      0.86      0.84        98
           1       0.76      0.70      0.73       110
           2       0.71      0.84      0.77        86
           3       0.96      0.88      0.92       106

    accuracy                           0.81       400
   macro avg       0.82      0.82      0.82       400
weighted avg       0.82      0.81      0.82       400


================ MATRIZ DE CONFUSÃO ================
[[84 14  0  0]
 [17 77 16  0]
 [ 0 10 72  4]
 [ 0  0 13 93]]


### RESULTADOS COM K=1

```text
Total de testes: {400}
Total de acertos: {288}
Acurácia: {0.72}
Acurácia em porcentagem: {72.0} %


### RESULTADOS COM K=5

```text
Total de testes: {400}
Total de acertos: {313}
Acurácia: {0.7825}
Acurácia em porcentagem: {78.25} %


### RESULTADOS COM K=10

```text
Total de testes: {400}
Total de acertos: {316}
Acurácia: {0.79}
Acurácia em porcentagem: {79.0} %

### RESULTADOS COM K=15

```text
Total de testes: {400}
Total de acertos: {324}
Acurácia: {0.81}
Acurácia em porcentagem: {81.0} %

### RESULTADOS COM K=20

```text
Total de testes: {400}
Total de acertos: {325}
Acurácia: {0.8125}
Acurácia em porcentagem: {81.25} %

### RESULTADOS COM K=30

```text
Total de testes: {400}
Total de acertos: {326}
Acurácia: {0.815}
Acurácia em porcentagem: {81.50} %

### RESULTADOS COM K=50

```text
Total de testes: {400}
Total de acertos: {328}
Acurácia: {0.82}
Acurácia em porcentagem: {82.0} %

### RESULTADOS COM K=60

```text
Total de testes: {400}
Total de acertos: {329}
Acurácia: {0.8225}
Acurácia em porcentagem: {82.25} %

### RESULTADOS COM K=70

```text
Total de testes: {400}
Total de acertos: {322}
Acurácia: {0.805}
Acurácia em porcentagem: {80.50} %

### RESULTADOS COM K=250

```text
Total de testes: {400}
Total de acertos: {305}
Acurácia: {0.7625}
Acurácia em porcentagem: {76.25} %

### Interpretando os valores de K

```text
Iniciando os testes com K=1, obtivemos uma acurácia de 72%.
Já com K=5, tivemos um aumento significativo da acurária em 7.25%, chegando em 78.25%.
Com valores de K entre 10 e 60, tivemos continuamos obtendo um aumento na acurácia, chegando em 82.25% com K=60.
Ao aumentarmos o K para 70, a acurácia começou a cair, chegando em 80.50% para K=70.
Com valor de K=250 , a acurácia fica em 76.25%.
Com isso, concluímos que nosso valor de K ideal para esse treinamento seria K=60, onde obtivemos o melhor resultado!
